System/Random.hs

```diff
@@ -1071,24 +1071,75 @@ randomIvalInteger (l,h) rng
                         (x,g') = next g
                         v' = (v * b + (fromIntegral x - fromIntegral genlo))
 
+-- | Generate an 'Integer' in the range @[l, h]@ if @l <= h@ and @[h, l]@
+-- otherwise.
 uniformIntegerM :: (MonadRandom g m) => (Integer, Integer) -> g -> m Integer
-uniformIntegerM (l, h) gen
-  | l > h = uniformIntegerM (h, l) gen
-  | otherwise = do
-    v <- f 1 0
-    pure (l + v `mod` k)
+uniformIntegerM (l, h) gen = case l `compare` h of
+  LT -> do
+    let limit = h - l
+    let limitAsWord64 :: Word64 = fromIntegral limit
+    bounded <-
+      if (toInteger limitAsWord64) == limit
```

Comment:
Thanks for the suggestions!

> At the same time, it is not fair that negative values don't have that optimization.

Which negative values? `limit` is always positive here because it is the difference between `h` and `l` with `h > l`.

> So I recommend checking the `Int` instead of `Word64` (Int is a better choice because then it will have a direct conversion to `S# Int#`):
> 
> ```haskell
> LT | l >= toInteger (minBound :: Int) && h <= toInteger (maxBound :: Int) -> toInteger <$> uniformR (fromIntegral l, fromIntegral h) gen
>    | otherwise -> boundedExclusiveIntegerM (h - l + 1) gen
> ```

In fact, the approach as it stands optimises more cases. E.g. if both `l` and `h` a very large but not far apart, then with the current approach the optimisation kicks in: say `l = 10^100` and `h = 10^100 + 2^60`. Then their difference fits into a `Word64` and we can use the fast path.

This does not work if both `l` and `h` need to fit into `Int` as in the alternative proposal.

I benchmarked the alternative proposal and it is slower than the current approach for "small" Integers (i.e. ones that fit into `Int`):

```
      1,263,565 randoms generated [System.Random Integers]    ~ 1,071 cycles/int (vs. 494)
        499,462 randoms generated [System.Random Integers]    ~ 2,711 cycles/int (vs. 1,130)
```

Reply:
What I mean by

> At the same time, it is not fair that negative values don't have that optimization.

is that if I ask for `uniformR (-100, 100)`, I will not get this optimization, because it only works on positive numbers. 

Not sure why the performance is worse of what I suggested, but what I am trying to say is that
 negative ranges are just as important, but it does not have implemented in the way I wrote it.

Reply:
> I ask for `uniformR (-100, 100)`, I will not get this optimization, because it only works on positive numbers.

```haskell
uniformR (-100 :: Integer, 100 :: Integer)
```
calls
```haskell
uniformIntegerM (-100 :: Integer, 100 :: Integer)
```
and we get
```haskell
l :: Integer = -100
h :: Integer = 100
limit :: Integer = 100 - (-100)
limitAsWord64 :: Word64 = fromIntegral limit
```

with `toInteger limitAsWord64 == limit`, and so the optimisation is triggered for this call.

Reply:
Awesome. In that case disregard everything that I said ;)